File: Sources/CoreGraphicsKitExtras/Geometric Data Types/CoreGraphicsRectangle.swift
//===----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------===//
//
//  CoreGraphicsRectangle.swift
//  core-graphics-kit
//
//  Created by Fang Ling on 2026/8/8.
//
//  This source file is part of the CoreGraphicsKit open source project
//
//  Copyright (c) 2026 Fang Ling <fangling@fangl.ing>
//  Licensed under Apache License v2.0
//
//  See LICENSE for license information
//
//  SPDX-License-Identifier: Apache-2.0
//
//===----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------===//

import CKit
import CoreGraphicsKitEssentials

extension CoreGraphicsRectangle {
  /// Creates a rectangle with the specified coordinate and size values.
  ///
  /// - Parameters:
  ///   - x: The x-coordinate of the rectangle's origin point.
  ///   - y: The y-coordinate of the rectangle's origin point.
  ///   - width: The width of the rectangle.
  ///   - height: The height of the rectangle.
  public init(x: CFloatingPoint64, y: CFloatingPoint64, width: CFloatingPoint64, height: CFloatingPoint64) {
    self.init(origin: CoreGraphicsPoint(x: x, y: y), size: CoreGraphicsSize(width: width, height: height))
  }
}

extension CoreGraphicsRectangle: @retroactive Swift::Equatable {
  public static func == (lhs: Self, rhs: Self) -> CBoolean {
    return CoreGraphicsRectangleIsEqual(lhs, rhs)
  }
}

File: Sources/CoreGraphicsKitEssentials/Geometric Data Types/CoreGraphicsRectangle.c
/*===----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------===*
 *
 *  CoreGraphicsRectangle.c
 *  core-graphics-kit
 *
 *  Created by Fang Ling on 2026/8/2.
 *
 *  This source file is part of the CoreGraphicsKit open source project
 *
 *  Copyright (c) 2026 Fang Ling <fangling@fangl.ing>
 *  Licensed under Apache License v2.0
 *
 *  See LICENSE for license information
 *
 *  SPDX-License-Identifier: Apache-2.0
 *
 *===----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------===*/

#include "CoreGraphicsRectangle.h"

C_ASSUME_NONNULL_BEGIN

CBoolean CoreGraphicsRectangleIsEqual(CoreGraphicsRectangle r1, CoreGraphicsRectangle r2) {
  return CoreGraphicsPointIsEqual(r1.origin, r2.origin) && CoreGraphicsSizeIsEqual(r1.size, r2.size);
}

C_ASSUME_NONNULL_END
